pkg/storage/etcd/v3_helper.go

```diff
@@ -0,0 +1,216 @@
+	if ctx == nil {
+		glog.Errorf("Context is nil")
+		ctx = context.TODO()
+	}
+
+	rev, err := extractRevision(h.versioner, obj)
+	if err != nil {
+		return err
+	}
+
+	data, err := runtime.Encode(h.codec, obj)
+	if err != nil {
+		return err
+	}
+	key = prefixEtcdKey(key, h.pathPrefix)
+
+	// Note: if rev = 0, it is equivalent to create.
+	txnResp, err := h.client.KV.Txn(ctx).If(
+		clientv3.Compare(clientv3.ModifiedRevision(key), "=", rev),
+	).Then(
+		clientv3.OpPut(key, string(data)),
+	).Commit()
+	if err != nil {
+		return err
+	}
+	if !txnResp.Succeeded {
+		return fmt.Errorf("Set: compare failed. Key: %s, Revision: %d", key, rev)
+	}
+
+	if out != nil {
+		// get the response to the put request
+		putResp := txnResp.Responses[0].GetResponsePut()
+		if _, err := conversion.EnforcePtr(out); err != nil {
+			panic("unable to convert output object to pointer")
+		}
+		return decode(h.codec, h.versioner, data, out, putResp.Header.Revision)
+	}
+
+	return nil
+}
+
+func (h *v3Helper) Get(ctx context.Context, key string, out runtime.Object, ignoreNotFound bool) error {
+	if ctx == nil {
+		glog.Errorf("Context is nil")
+		ctx = context.TODO()
+	}
+	key = prefixEtcdKey(key, h.pathPrefix)
+	getResp, err := h.client.KV.Get(ctx, key)
+	if err != nil {
+		return err
+	}
+
+	// if not found:
+	//   if ignoreNotFound: set objPtr to be zero value
+	//   else: return not found error
+	if len(getResp.Kvs) == 0 {
+		if ignoreNotFound {
+			return setZeroValue(out)
+		} else {
+			return ErrNotFound
+		}
+	}
+
+	kv := getResp.Kvs[0]
+	return decode(h.codec, h.versioner, kv.Value, out, kv.ModRevision)
+}
+
+func (h *v3Helper) Create(ctx context.Context, key string, obj, out runtime.Object, ttl uint64) error {
+	panic("unimplemented")
+}
+
+func (h *v3Helper) Delete(ctx context.Context, key string, out runtime.Object) error {
+	panic("unimplemented")
+}
+
+func (h *v3Helper) GetToList(ctx context.Context, key string, filter storage.FilterFunc, listObj runtime.Object) error {
+	panic("unimplemented")
+}
+
+func (h *v3Helper) List(ctx context.Context, key, resourceVersion string, filter storage.FilterFunc, listObj runtime.Object) error {
+	panic("unimplemented")
+}
+
+func (h *v3Helper) GuaranteedUpdate(ctx context.Context, key string, ptrToType runtime.Object, ignoreNotFound bool, tryUpdate storage.UpdateFunc) error {
+	panic("unimplemented")
+}
+
+func (h *v3Helper) Watch(ctx context.Context, key string, resourceVersion string, filter storage.FilterFunc) (watch.Interface, error) {
+	panic("unimplemented")
+}
+
+func (h *v3Helper) WatchList(ctx context.Context, key string, resourceVersion string, filter storage.FilterFunc) (watch.Interface, error) {
+	panic("unimplemented")
+}
+
+func prefixEtcdKey(key, prefix string) string {
+	if strings.HasPrefix(key, prefix) {
+		return key
+	}
+	return path.Join(prefix, key)
+}
+
+func extractRevision(versioner storage.Versioner, obj runtime.Object) (int64, error) {
```

Comment:
I think Tim was talking in this context:
https://github.com/hongchaodeng/kubernetes/commit/21f089f27f94440ce6713630168f368c411cfaa1#commitcomment-16690709